packages/connection-storage/src/connection-storage.ts

```diff
@@ -1,67 +1,123 @@
-    const fetchConnectionModels = promisifyAmpersandMethod(
-      connectionCollection.fetch.bind(connectionCollection)
+  private readonly folder = 'Connections';
+  private readonly keytarServiceName: string;
+
+  constructor(protected readonly path: string = '') {
+    this.keytarServiceName = getKeytarServiceName();
+  }
+
+  private getFolderPath() {
+    return join(this.path, this.folder);
+  }
+
+  private getFilePath(id: string) {
+    return join(this.getFolderPath(), `${id}.json`);
+  }
+
+  private mapStoredConnectionToConnectionInfo(
+    storedConnectionInfo: ConnectionInfo,
+    secrets?: ConnectionSecrets
+  ): ConnectionInfo {
+    const connectionInfo = mergeSecrets(storedConnectionInfo, secrets);
+    if (connectionInfo.lastUsed) {
+      connectionInfo.lastUsed = new Date(connectionInfo.lastUsed);
+    }
+    return deleteCompassAppNameParam(connectionInfo);
+  }
+
+  private async getKeytarCredentials() {
+    const credentials = await keytar.findCredentials(this.keytarServiceName);
+    return Object.fromEntries(
+      credentials.map(({ account, password }) => [
+        account,
+        JSON.parse(password).secrets ?? {},
+      ])
+    ) as Record<string, ConnectionSecrets>;
+  }
+
+  private async getConnections(): Promise<any[]> {
+    const connectionIds = (await fs.readdir(this.getFolderPath()))
+      .filter((file) => file.endsWith('.json'))
+      .map((file) => file.replace('.json', ''));
+
+    const data = await Promise.all(
+      connectionIds.map(async (id) => {
+        try {
+          return JSON.parse(await fs.readFile(this.getFilePath(id), 'utf8'));
+        } catch (e) {
+          log.error(
+            mongoLogId(1_001_000_200),
+            'Connection Storage',
+            'Failed to parse connection',
+            { message: (e as Error).message, connectionId: id }
+          );
+          return undefined;
+        }
+      })
     );
+    return data.filter(Boolean);
+  }
+
+  async hasLegacyConnections() {
+    return (
+      (await this.getConnections()).filter((x) => !x.connectionInfo).length > 0
+    );
+  }
+
+  async loadAll(): Promise<ConnectionInfo[]> {
+    // Ensure folder exists
+    await fs.mkdir(this.getFolderPath(), { recursive: true });
 
     try {
-      await fetchConnectionModels();
+      const connections = (await this.getConnections())
+        // Ignore legacy connections and make sure connection has a connection string.
+        .filter(
+          (x: { connectionInfo?: ConnectionInfo }) =>
+            x.connectionInfo &&
+            x.connectionInfo.connectionOptions.connectionString
+        )
+        .map((x) => x.connectionInfo) as ConnectionInfo[];
+
+      if (process.env.COMPASS_E2E_DISABLE_KEYCHAIN_USAGE === 'true') {
+        return connections.map((connection) =>
+          this.mapStoredConnectionToConnectionInfo(connection)
+        );
+      }
+
+      const secrets = await this.getKeytarCredentials();
```

Comment:
Probably not a big deal, but does it make sense to perform `this.getKeytarCredentials()` and the `this.getConnections()` call above in parallel?

Reply:
that's a good idea.